src/Microsoft.AspNetCore.Mvc.Core/Controllers/ControllerFeatureProvider.cs

```diff
@@ -0,0 +1,79 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using Microsoft.AspNetCore.Mvc.ApplicationParts;
+
+namespace Microsoft.AspNetCore.Mvc.Controllers
+{
+    /// <summary>
+    /// Discovers controllers from a list of <see cref="ApplicationPart"/> instances.
+    /// </summary>
+    public class ControllerFeatureProvider : IApplicationFeatureProvider<ControllerFeature>
+    {
+        private const string ControllerTypeNameSuffix = "Controller";
+
+        /// <inheritdoc />
+        public void PopulateFeature(
+            IEnumerable<ApplicationPart> parts,
+            ControllerFeature feature)
+        {
+            foreach (var part in parts.OfType<IApplicationPartTypeProvider>())
+            {
+                foreach (var type in part.Types)
+                {
+                    if (IsController(type) && !feature.Controllers.Contains(type))
```

Comment:
Isn't this all speculative perf-doing? (spoiler: speculative hand-waving about perf will *not* be part of my 10 top tips)





*There's also a simple solution here that resolves the perf concern and is probably more correct than what the code is doing.*